File: src/fairshare/splitter.py
"""Split an expense amount (in cents) among participants."""

from __future__ import annotations

from fairshare.errors import ValidationError


def equal_split(amount_cents: int, n: int) -> list[int]:
    """Divide *amount_cents* equally among *n* participants.

    Uses integer division; distributes remainder cents one-by-one to the
    first ``remainder`` participants so the total always sums to *amount_cents*.

    Raises :class:`ValidationError` for non-positive amounts or n < 1.
    """
    if amount_cents <= 0:
        raise ValidationError(f"amount_cents must be positive, got {amount_cents}")
    if n < 1:
        raise ValidationError(f"n must be at least 1, got {n}")

    base = amount_cents // n
    remainder = amount_cents % n
    return [base + (1 if i < remainder else 0) for i in range(n)]


def weighted_split(amount_cents: int, weights: list[int]) -> list[int]:
    """Split *amount_cents* proportionally by *weights* (Hamilton / largest-remainder method).

    Each share uses integer division: ``amount_cents * weights[i] // total_weight``.
    Remaining cents go to the largest remainders, then original index.

    Raises :class:`ValidationError` for non-positive amounts, empty weights,
    or non-positive weights.
    """
    if amount_cents <= 0:
        raise ValidationError(f"amount_cents must be positive, got {amount_cents}")
    if not weights:
        raise ValidationError("weights must not be empty")
    if any(w <= 0 for w in weights):
        raise ValidationError(f"All weights must be positive, got {weights}")

    total_weight = sum(weights)
    floors = [amount_cents * w // total_weight for w in weights]
    remainders = [amount_cents * w % total_weight for w in weights]
    leftover = amount_cents - sum(floors)

    # Hamilton / largest-remainder: leftover cents go to highest remainder, then index
    order = sorted(range(len(weights)), key=lambda i: (-remainders[i], i))
    for i in range(leftover):
        floors[order[i]] += 1

    return floors

File: packages/domain/src/splitter.ts
import { ValidationError } from "./errors";

export function equalSplit(amountCents: number, n: number): number[] {
  if (amountCents <= 0) {
    throw new ValidationError(`amount_cents must be positive, got ${amountCents}`);
  }
  if (n < 1) {
    throw new ValidationError(`n must be at least 1, got ${n}`);
  }

  const base = Math.floor(amountCents / n);
  const remainder = amountCents % n;
  return Array.from({ length: n }, (_, i) => base + (i < remainder ? 1 : 0));
}

export function weightedSplit(amountCents: number, weights: number[]): number[] {
  if (amountCents <= 0) {
    throw new ValidationError(`amount_cents must be positive, got ${amountCents}`);
  }
  if (weights.length === 0) {
    throw new ValidationError("weights must not be empty");
  }
  if (weights.some((w) => w <= 0)) {
    throw new ValidationError(`All weights must be positive, got ${weights}`);
  }

  const totalWeight = weights.reduce((a, b) => a + b, 0);
  const floors = weights.map((w) => Math.floor((amountCents * w) / totalWeight));
  const remainders = weights.map((w) => (amountCents * w) % totalWeight);
  const leftover = amountCents - floors.reduce((a, b) => a + b, 0);

  const order = [...weights.keys()].sort((i, j) => remainders[j] - remainders[i] || i - j);
  for (let k = 0; k < leftover; k++) {
    floors[order[k]] += 1;
  }

  return floors;
}
